src/main/java/net/fabricmc/loader/util/DefaultLanguageAdapter.java

```diff
@@ -50,78 +54,98 @@ private DefaultLanguageAdapter() {
+				throw new LanguageAdapterException("Ambiguous " + value + " - refers to both field and method!");
 			}
 
-			if (methodList.isEmpty()) {
-				throw new LanguageAdapterException("Could not find " + value + "!");
-			} else if (methodList.size() >= 2) {
-				throw new LanguageAdapterException("Found multiple method entries of name " + value + "!");
+			if (!type.isAssignableFrom(fType)) {
+				throw new LanguageAdapterException("Field " + value + " cannot be cast to " + type.getName() + "!");
 			}
 
-			final Method targetMethod = methodList.get(0);
-			Object object = null;
+			//noinspection unchecked
+			return (T) field.get(null);
+		} catch (NoSuchFieldException e) {
+			// ignore
+		} catch (IllegalAccessException e) {
+			throw new LanguageAdapterException("Field " + value + " cannot be accessed!", e);
+		}
 
-			if ((targetMethod.getModifiers() & Modifier.STATIC) == 0) {
-				try {
-					object = c.getDeclaredConstructor().newInstance();
-				} catch (Exception e) {
-					throw new LanguageAdapterException(e);
-				}
-			}
+		if (foundMethod == null) {
+			throw new LanguageAdapterException("Could not find " + value + "!");
+		}
 
-			final Object targetObject = object;
+		if (!type.isInterface()) {
+			throw new LanguageAdapterException("Cannot proxy method " + value + " to non-interface type " + type.getName() + "!");
+		}
 
-			//noinspection unchecked
-			return (T) Proxy.newProxyInstance(FabricLauncherBase.getLauncher().getTargetClassLoader(), new Class[] { type }, new InvocationHandler() {
-				@Override
-				public Object invoke(Object proxy, Method method, Object[] args) throws Throwable {
-					return targetMethod.invoke(targetObject, args);
-				}
-			});
+		MethodHandles.Lookup lookup = MethodHandles.lookup();
+		MethodHandle called;
+		try {
+			called = lookup.unreflect(foundMethod);
+		} catch (IllegalAccessException ex) {
+			throw new LanguageAdapterException("Method " + value + " is not accessible to the language adapter", ex);
+		}
+
+		if (!Modifier.isStatic(foundMethod.getModifiers())) {
+			try {
+				called = called.bindTo(lookup.findConstructor(c, MethodType.methodType(Void.TYPE)).invoke());
+			} catch (Throwable e) {
+				throw new LanguageAdapterException("Cannot create a class instance to invoke instance method for " + value, e);
+			}
 		}
+
+		final MethodHandle target = called;
+		// cannot use lambda meta factory because hidden classes will not work
+		//noinspection unchecked
+		return (T) Proxy.newProxyInstance(FabricLauncherBase.getLauncher().getTargetClassLoader(), new Class[] {type}, (proxy, method, args) -> {
+			if (method.getDeclaringClass() == Object.class) {
+				switch (method.getName()) {
+				case "hashCode":
+					return System.identityHashCode(proxy);
+				case "equals":
+					return (proxy == args[0] ? Boolean.TRUE : Boolean.FALSE);
+				case "toString":
+					return proxy.getClass().getName() + '@' + Integer.toHexString(proxy.hashCode());
+				}
+			}
+
+			try {
```

Comment:
If we have moved to java 16, we need to call this: https://docs.oracle.com/en/java/javase/16/docs/api/java.base/java/lang/reflect/InvocationHandler.html#invokeDefault(java.lang.Object,java.lang.reflect.Method,java.lang.Object...)

Reply:
Turns out we always had this thing: https://docs.oracle.com/en/java/javase/16/docs/api/java.base/java/lang/invoke/MethodHandleProxies.html#asInterfaceInstance(java.lang.Class,java.lang.invoke.MethodHandle)

We probably... should be using that, *available since Java 7*, instead? After all, I am bad at Java:tm:!

This API also supports default methods since Java 12, so should be a good alternative.

Reply:
Just discovered that the bug was fixed in jdk 8 updates as well, so just wait for me to fix it...